goalc/compiler/compilation/Type.cpp

```diff
@@ -85,6 +85,75 @@ RegVal* Compiler::compile_get_method_of_object(RegVal* object,
   return deref->to_reg(env);
 }
 
+Val* Compiler::compile_format_string(const goos::Object& form,
+                                     Env* env,
+                                     std::string& fmt_template,
+                                     std::vector<RegVal*> args,
+                                     const std::string& out_stream) {
+  // Add first two format args
+  args.insert(args.begin(), compile_string(fmt_template, env, DEBUG_SEGMENT)->to_reg(env));
+  args.insert(args.begin(), compile_get_sym_obj(out_stream, env)->to_reg(env));
+
+  // generate code in the method_env
+  auto format_function = compile_get_symbol_value("_format", env)->to_gpr(env);
+  return compile_real_function_call(form, format_function, args, env);
+}
+
+Val* Compiler::generate_inspector_for_type(const goos::Object& form, Env* env, Type* type) {
+  auto as_structure_type = dynamic_cast<StructureType*>(type);
+  if (!as_structure_type) {  // generate the inspect method
+    return get_none();
+  }
+
+  StructureType* structured_type = as_structure_type;
+
+  // Create a function environment to hold the code for the inspect method. The name is just for
+  // debugging.
+  auto method_env = std::make_unique<FunctionEnv>(env, "autogenerated-inspect-method");
+  // put the method in the debug segment.
+  method_env->set_segment(DEBUG_SEGMENT);
+
+  // Create a register which will hold the input to the inspect method
+  auto input = method_env->make_ireg(structured_type->get_name(), emitter::RegKind::GPR);
+  // "Constraint" this register to be the register that the function argument is passed in
+  IRegConstraint constraint;
+  constraint.instr_idx = 0;         // constraint at the start of the function
+  constraint.ireg = input->ireg();  // constrain this register
+  constraint.desired_register = emitter::gRegInfo.get_arg_reg(0);  // to the first argument
+  method_env->constrain(constraint);
+
+  std::string str_template = "";
+  std::vector<RegVal*> formatArgs = {};
+
+  // Check if there are no fields
+  if (structured_type->fields().empty()) {
+    str_template = "Type has no fields!~%";
+  } else {
+    for (Field f : structured_type->fields()) {
+      str_template += "~A~%";
```

Comment:
I'm imagining that this could work a little bit differently.  You could put the field's name directly in the `str_template`, then print out the actual values of the fields of the object that's passed to the inspect.

So for a field named `name`, the `str_template` would be something like `name: ~A`, and the you would add the value of `name` field to the `formatArgs`.  I'll leave more details in another comment.